Sprint 7 solution in file manager

AFil0n opened the pull request.

AFil0n committed:
Добавлен класс FileBackedTaskManager

AFil0n committed:
Добавлено наследование FileBackedTaskManager  от InMemoryTaskManager

AFil0n committed:
Добавлен метод сохранения и доработаны методы добавления задач

AFil0n committed:
Доработаны методы toString у базовых реализаций задач

AFil0n committed:
Добаввлены методы fromString для реализаций тасков

AFil0n committed:
Добавлено исключение ManagerSaveException

AFil0n committed:
Доработан метод сохрания задач в файл

AFil0n committed:
Переписаны методы получения задач из строк

AFil0n committed:
Добавлен Enum типов тасков

AFil0n committed:
Переписаны методы сохранения задач

AFil0n committed:
Добавлен метод load для загрузки данных из файла,  и  проведен небольшой рефакторинг

AFil0n committed:
Добавлено обновление файла, при удалении задач

AFil0n committed:
Добавлено сохранение при изменении задач

AFil0n committed:
Изменен стандартный менеджер работы с задачами

AFil0n committed:
Добавлены конструкторы FileBackedTaskManager

AFil0n committed:
Добавлен файл taskc.csv в игнор лист

AFil0n committed:
Исправлены ошибки записи в файл

AFil0n committed:
Поправлен кодстайл всего проекта

AFil0n committed:
Исправлен автотест TaskManagerTest

AFil0n committed:
Исправление ошибок получения данных и сохранения в файл

AFil0n committed:
Исправлена ошибка получения данных из фала

AFil0n committed:
Добавлено тестировани добавления тасков из файла

AFil0n committed:
Добавлено сохранение в файл при удалении задачи

AFil0n committed:
Добавлен тест удаления таска из файла

AFil0n committed:
Обновление типов задач в при переводе в строку

AFil0n committed:
Добавлено обновление эпика при добавлении сабтасков с файла

AFil0n committed:
Добавлен тест на создание эпиков и сабтасков

AFil0n committed:
Доработано форматирование

AFil0n committed:
Удалил путые строки

AFil0n committed:
Добавлен тест на удаление сабтасков

AFil0n committed:
Добавлена проверка на удаление епиков

AFil0n committed:
Переписаны тесты

AFil0n committed:
Убрал неиспользуемые методы

AFil0n committed:
Фикс кодстайла

AFil0n committed:
Кодстайл фикс

AFil0n committed:
поправлен кодстайл

AFil0n committed:
Исправлены тесты

sergeymamonov commented on src/com/yandex/app/model/Epic.java:
```diff
@@ -32,6 +43,18 @@ public void cleanSubtasksIds() {
+    public static Task fromString(String taskString) {
+        String[] data = taskString.split(",");
+
+        if (data.length != 5) {
```
Здесь 5 - это магическая константа. Из кода непонятно, что значит это число. Его нужно вынести в константу с информативным названием.

sergeymamonov commented on src/com/yandex/app/model/Subtask.java:
```diff
@@ -16,4 +22,20 @@ public Subtask(String name, String description, int epicId) {
+    public static Task fromString(String taskString) {
+        String[] data = taskString.split(",");
+
+        if (data.length != 6) {
```
Аналогично.

sergeymamonov commented on src/com/yandex/app/model/Task.java:
```diff
@@ -49,8 +58,33 @@ public void setStatus(Status status) {
+    public static Task fromString(String taskString) {
+        String[] data = taskString.split(",");
+
+        if (data.length != 5) {
```
И еще раз.

sergeymamonov commented on src/com/yandex/app/service/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,201 @@
+            try {
+                file.createNewFile();
+            } catch (IOException e) {
+                throw new RuntimeException(e);
```
Не нужно кидать RuntimeException. Нужно создать свое исключение и кидать его. ManagerSaveException, указано в задании.

sergeymamonov commented on src/com/yandex/app/service/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,201 @@
+                    writer.newLine();
+                }
+            }
+        } catch (ManagerSaveException e) {
```
Какой метод может кинуть ManagerSaveException?
И не нужно ниже кидать RuntimeException.
Одного catch достаточно.
Нужно поймать IOException и кинуть ManagerSaveException.

sergeymamonov commented on src/com/yandex/app/service/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,201 @@
+    }
+
+    public void loadFromFile(File file) {
+        final TaskManager taskManager = new InMemoryTaskManager();
```
Переменная нигде не используется. Обращайте внимание на подсказки среды разработки.

sergeymamonov commented on src/com/yandex/app/service/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,201 @@
+                    default -> putSubTask(line);
+                }
+            }
+        } catch (NullPointerException e) {
```
Мы не ловим NPE. NPE это ошибка разработчика. И если она где-то и возникает, то нужно править код.

sergeymamonov commented on src/com/yandex/app/service/InMemoryTaskManager.java:
```diff
@@ -238,4 +240,8 @@ private void updateEpicStatus(int epicId) {
     }
+
+    void setCounterMaxId(int id) {
+        counter = id > counter ? id : counter;
```
Используйте стандартный метод max(), не нужно писать методы, которые можно найти уже написанные.

sergeymamonov commented on src/com/yandex/app/service/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,201 @@
+package com.yandex.app.service;
+
+import com.yandex.app.model.*;
```
Лучше не использовать импорт через *, так как это может приводить к конфликтам импорта из разных пакетов. Если это выполняется автоматически идеей, то лучше отключить wildcard import https://www.jetbrains.com/help/idea/creating-and-optimizing-imports.html#disable-wildcards-for-class.

AFil0n committed:
Вынес количество элементов для парсинга в переменую

AFil0n committed:
Добавил в игнор настройки idea

AFil0n committed:
Убрано использование импорта через *

AFil0n committed:
Изменено исключение с RuntimeException на ManagerSaveException

AFil0n committed:
Замена выкидывания RuntimeException на ManagerSaveException

AFil0n committed:
Убрана не используемая переменая

AFil0n committed:
Убран отлов NullPointerException

AFil0n committed:
Изменено самописное получение максимального таска на получение таска по

Изменено самописное получение максимального таска на получение таска получение максимального таска через Math.max

AFil0n committed:
Исправлен импорт через * на импорт, каждого класса по отдельности

AFil0n committed:
Исправлена ошибка приводящая к NullPointerExeption

AFil0n committed:
Перенес парамтер parsingParamsCount в метод fromString, так как используется только в нем и другое использование на данный момент не предпологается

sergeymamonov commented on src/com/yandex/app/model/Epic.java:
```diff
@@ -32,6 +43,19 @@ public void cleanSubtasksIds() {
+
+    public static Task fromString(String taskString) {
+        String[] data = taskString.split(",");
+        final int parsingParamsCount = 5;
```
Лучше выделить в константу private static final ...

sergeymamonov commented on src/com/yandex/app/service/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,204 @@
+                }
+            }
+        }  catch (IOException e) {
+            throw new RuntimeException("Ошибка в файле: " + file.getAbsolutePath(), e);
```
Разработчик не кидает RuntimeException. Для этого в Java есть механизм наследования. И свою ошибку можно унаследовать от нужной, например от RuntimeException. Оно RuntimeException разработчик не кидает.
В задании указано какую ошибку нужно кинуть.

sergeymamonov commented on src/com/yandex/app/service/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,204 @@
+import com.yandex.app.model.TaskType;
+import com.yandex.app.util.ManagerSaveException;
+
+import java.io.*;
```
Воспользуйтесь поиском по всему проекту и поищите звездочку с точкой запятой.

sergeymamonov commented on src/com/yandex/app/model/Task.java:
```diff
@@ -49,8 +58,34 @@ public void setStatus(Status status) {
+
+    public static Task fromString(String taskString) {
+        String[] data = taskString.split(",");
+        final int parsingParamsCount = 5;
```
Просмотрите весь код на наличие магических констант и выделите их в константы с информативными названиями.

sergeymamonov commented on src/com/yandex/app/service/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,204 @@
+                switch (type) {
+                    case TASK -> putTask(line);
+                    case EPIC -> putEpic(line);
+                    default -> putSubTask(line);
```
Не нужно так делать. Завтра появится новый тип зада. Здесь вы забудете добавить этот новый тип. И в putSubTask - вы пойдете и для сабтасок и для нового типа задач. И скорее всего для нового типа будет ошибка. Поэтому для каждого типа задач свой кейс, а в default чаще всего кидают ошибку о том что получен неподдерживаемый тип.
